src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3451,12 +3451,34 @@ bool os::pd_create_stack_guard_pages(char* addr, size_t size) {
                                                            (size_t)addr - stack_extent);
     }
 
-    if (stack_extent < (uintptr_t)addr) {
-      ::munmap((void*)stack_extent, (uintptr_t)(addr - stack_extent));
+    // stack grows towards lower address
+    //
+    // lower address first
+    //
+    // |             |
+    // |    stack    |
+    // |    frame    |
+    // |             | <----- stack_extent (stack end)
+    // |             |
+    // |xxxxxxxxxxxxx| <------ addr
+    // |xxxxxxxxxxxxx|
+    // |xxxxxxxxxxxxx|         stack-guard-pages
+    // |xxxxxxxxxxxxx| <------ addr + size
+    // |             |
+    // |             |
+    // |ooooooooooooo| <------ actually used by thread
+    // |ooooooooooooo|
+    // |ooooooooooooo|
+    // |ooooooooooooo| <------ stack start
+    //
+    // Truncate the mapping if there is overlapping.
+    if (stack_extent < (uintptr_t)addr + size) {
+      ::munmap((void*)stack_extent, (uintptr_t)(addr + size - stack_extent));
     }
+    return os::pd_attempt_reserve_memory_at(addr, size, !ExecMem) != nullptr;
```

Comment:
This patch indeed goes beyond my original intention. IIRC,  my intention was to replace `os::commit()` with `::mprotect()` to set protection on stack guard pages.

Reply:
> Current create_stack_guard_pages() implementation calls os::commit_memory(addr, size, !ExecMem) to setup the protection.

Which `create_stack_guard_pages` are you referring to, in `StackOverflow` or `os`?

`StackOverflow::create_stack_guard_pages` invokes `os::create_stack_guard_pages` for some preparation work (optional for some platforms) and `os::guard_memory` for protection setup.

`os::commit_memory` is not directly related to memory protection.

Reply:
> > Current create_stack_guard_pages() implementation calls os::commit_memory(addr, size, !ExecMem) to setup the protection.
> 
> Which `create_stack_guard_pages` are you referring to, in `StackOverflow` or `os`?
> 
> `StackOverflow::create_stack_guard_pages` invokes `os::create_stack_guard_pages` for some preparation work (optional for some platforms) and `os::guard_memory` for protection setup.
> 
> `os::commit_memory` is not directly related to memory protection.

Ah, okay.  I believe I wanted to get rid of this `commit_memory` call which has the undesirable side-effort to NMT. 
BTW, do you need the `reserve_memory()`, are the guard pages already reserved as part of stack?